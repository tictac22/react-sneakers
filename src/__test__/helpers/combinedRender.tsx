import { render } from "@testing-library/react"
import { Provider } from "react-redux"
import { MemoryRouter } from "react-router"
import { createStore } from "../../redux/store"

type Params = {
	children: React.ReactNode
	store: {}
}
export const CombinedRender = ({ children, store }: Params) => {
	return render(
		<MemoryRouter>
			<Provider store={createStore(store)}>{children}</Provider>
		</MemoryRouter>
	)
}
