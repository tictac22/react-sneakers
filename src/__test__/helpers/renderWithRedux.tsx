import { render } from "@testing-library/react"
import { Provider } from "react-redux"
import { createStore } from "../../redux/store"

type Props = {
	children: React.ReactNode
	store: {}
}

export const renderWithRedux = ({ children, store }: Props) => {
	return render(<Provider store={createStore(store)}>{children}</Provider>)
}
