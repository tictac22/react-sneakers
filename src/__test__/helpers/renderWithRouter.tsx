import { render } from "@testing-library/react"
import { createMemoryHistory } from "history"
import { Router } from "react-router"

type Params = {
	children: React.ReactNode
}
export const renderWithRouter = ({ children }: Params) => {
	const history = createMemoryHistory()
	return {
		view: render(<Router history={history}>{children}</Router>),
		history,
	}
}
