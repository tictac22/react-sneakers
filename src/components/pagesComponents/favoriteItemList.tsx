import { Link } from "react-router-dom"
import { useAppSelector } from "../../redux/hooks"
import s from "../../styles/favorite.module.scss"
import { Cart } from "../cart"

export const FavoriteItemLists = () => {
	const { favoritesItems } = useAppSelector((state) => state.shop)

	if (favoritesItems.length === 0) return <EmptyFavorites />
	return (
		<div className="grid" style={{ justifyContent: "initial" }}>
			{favoritesItems.map((item) => (
				<Cart key={item.id} {...item} />
			))}
		</div>
	)
}

const EmptyFavorites = () => (
	<div className={s.empty}>
		<div className={s.empty__smile}>
			<img src="/images/smile.png" alt="smile" />
		</div>
		<h2 className={s.empty__title}>No bookmarks :(</h2>
		<p className={s.empty__text}>You haven't bookmarked anything</p>
		<Link to="/" className={`${s.empty__btn} btn`}>
			<img src="/svg/arrow.svg" />
			<p>Go back</p>
		</Link>
	</div>
)
