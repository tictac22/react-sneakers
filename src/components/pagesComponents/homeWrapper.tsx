import { useState } from "react"
import styles from "../../styles/home.module.scss"
import { Cart } from "../cart"
import { CartLoader } from "../cart/cartLoader"
import { useGetAllItems } from "../hooks/getItems"

export const HomeWrapper = () => {
	const [searchByname, setSearchByname] = useState<string>("")
	const { isError, isLoading, cartItems } = useGetAllItems()
	const filteredCartItem = cartItems.filter((item) =>
		item.title.toLowerCase().includes(searchByname.toLowerCase().trim())
	)
	return (
		<>
			<div className={styles.main__top}>
				<h1 className={styles.main__title}>
					{searchByname.trim() === "" ? "All sneakers" : `Result on request: '${searchByname}'`}
				</h1>
				<div className={styles.search}>
					<img src="/svg/search.svg" alt="search" />
					<input
						value={searchByname}
						maxLength={30}
						onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
							setSearchByname(e.target.value)
						}}
						className={`${styles.search__input}`}
						disabled={isError}
						placeholder="Searching..."
					/>
				</div>
			</div>
			<div className="grid">
				{isLoading ? (
					<CartLoader />
				) : isError ? (
					<p>Error</p>
				) : (
					filteredCartItem.map((item) => <Cart key={item.id} {...item} />)
				)}
			</div>{" "}
		</>
	)
}
