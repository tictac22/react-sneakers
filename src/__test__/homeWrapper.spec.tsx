import { screen, waitFor } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import { rest } from "msw"
import { setupServer } from "msw/node"
import { HomeWrapper } from "../components/pagesComponents/homeWrapper"
import { CombinedRender } from "./helpers/combinedRender"

const fakeData = [
	{
		title: "Nike Blazer Mid '77 Vintage Men's",
		price: 215,
		img: "img1",
		id: "3YuQN5SNO0ioX6adbAk3eQ",
	},
	{
		title: "Men's Nike Air Max 270",
		price: 260,
		img: "img2",
		id: "jRFlfF6IukWair2Jl_z",
	},
]
const handlers = [
	rest.get("https://631527a95b85ba9b11dcd427.mockapi.io/sneakers", (req, res, ctx) => {
		return res(ctx.json(fakeData))
	}),
]
export const server = setupServer(...handlers)
beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())
describe("Home Wrapper", () => {
	it("renders succesfully", async () => {
		const view = CombinedRender({ children: <HomeWrapper />, store: {} })
		expect(screen.getAllByLabelText("loading")[0]).toBeInTheDocument()
		await waitFor(() => screen.getByLabelText("Nike Blazer Mid '77 Vintage Men's"))

		expect(view.asFragment()).toMatchSnapshot()
	})

	it("fails render", async () => {
		server.use(
			rest.get("https://631527a95b85ba9b11dcd427.mockapi.io/sneakers", (req, res, ctx) => {
				return res(ctx.status(500))
			})
		)
		const view = CombinedRender({ children: <HomeWrapper />, store: {} })
		expect(screen.getAllByLabelText("loading")[0]).toBeInTheDocument()

		await waitFor(() => screen.getByLabelText("error"))

		expect(view.asFragment()).toMatchSnapshot()
	})
	it("filters by input", async () => {
		const view = CombinedRender({ children: <HomeWrapper />, store: {} })
		await waitFor(() => screen.getByLabelText("Men's Nike Air Max 270"))

		await userEvent.type(screen.getByPlaceholderText("Searching..."), "Nike Blazer")

		expect(screen.queryByLabelText("Men's Nike Air Max 270")).not.toBeInTheDocument()

		expect(view.asFragment()).toMatchSnapshot()
	})
})
