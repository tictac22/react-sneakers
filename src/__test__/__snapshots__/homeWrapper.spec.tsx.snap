// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Wrapper fails render 1`] = `
<DocumentFragment>
  <div>
    <h1>
      All sneakers
    </h1>
    <div>
      <img
        alt="search"
        src="/svg/search.svg"
      />
      <input
        class="undefined"
        disabled=""
        maxlength="30"
        placeholder="Searching..."
        value=""
      />
    </div>
  </div>
  <div
    class="grid"
  >
    <p
      aria-label="error"
    >
      Error
    </p>
  </div>
   
</DocumentFragment>
`;

exports[`Home Wrapper filters by input 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Result on request: 'Nike Blazer'
    </h1>
    <div>
      <img
        alt="search"
        src="/svg/search.svg"
      />
      <input
        class="undefined"
        maxlength="30"
        placeholder="Searching..."
        value="Nike Blazer"
      />
    </div>
  </div>
  <div
    class="grid"
  >
    <div
      aria-label="Nike Blazer Mid '77 Vintage Men's"
    >
      <div>
        <div
          aria-label="to favorite"
        >
          <img
            alt="unfavorite"
            src="/images/unfavorite.jpg"
          />
        </div>
        <div>
          <img
            alt="shoes"
            src="/images/shoes/img1.jpg"
          />
        </div>
        <h3>
          Nike Blazer Mid '77 Vintage Men's
        </h3>
        <div>
          <div>
            <div>
              Cost
            </div>
            <div>
              <span>
                $215
              </span>
            </div>
          </div>
          <div>
            <img
              alt="plus"
              src="/svg/plus.svg"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
   
</DocumentFragment>
`;

exports[`Home Wrapper renders succesfully 1`] = `
<DocumentFragment>
  <div>
    <h1>
      All sneakers
    </h1>
    <div>
      <img
        alt="search"
        src="/svg/search.svg"
      />
      <input
        class="undefined"
        maxlength="30"
        placeholder="Searching..."
        value=""
      />
    </div>
  </div>
  <div
    class="grid"
  >
    <div
      aria-label="Nike Blazer Mid '77 Vintage Men's"
    >
      <div>
        <div
          aria-label="to favorite"
        >
          <img
            alt="unfavorite"
            src="/images/unfavorite.jpg"
          />
        </div>
        <div>
          <img
            alt="shoes"
            src="/images/shoes/img1.jpg"
          />
        </div>
        <h3>
          Nike Blazer Mid '77 Vintage Men's
        </h3>
        <div>
          <div>
            <div>
              Cost
            </div>
            <div>
              <span>
                $215
              </span>
            </div>
          </div>
          <div>
            <img
              alt="plus"
              src="/svg/plus.svg"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      aria-label="Men's Nike Air Max 270"
    >
      <div>
        <div
          aria-label="to favorite"
        >
          <img
            alt="unfavorite"
            src="/images/unfavorite.jpg"
          />
        </div>
        <div>
          <img
            alt="shoes"
            src="/images/shoes/img2.jpg"
          />
        </div>
        <h3>
          Men's Nike Air Max 270
        </h3>
        <div>
          <div>
            <div>
              Cost
            </div>
            <div>
              <span>
                $260
              </span>
            </div>
          </div>
          <div>
            <img
              alt="plus"
              src="/svg/plus.svg"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
   
</DocumentFragment>
`;
